### DISCUSSION TITLE
Swift Concurrency Support

### DISCUSSION THREAD

**Discussion by fwpascual** [Tyler-Keith-Thompson/CucumberSwift#47] [Q&A] [answered]
Hello!  I was wondering what is the best way to use this library with Swift Concurrency async/await functions?  When testing in a standard XCTestCase, I can mark test and `setUp` functions with `async` to call into async functions, awaiting results as needed to verify the responses.

I can use `Task` from within a `Given` or `When` step, but I'm not sure how to handle situations where steps are dependent on previous steps completing.  I can likely come up with a custom solution but was wondering if there is an idiomatic way of handling this or if there are future plans to support Swift Concurrency within this framework itself?

**Comment by Tyler-Keith-Thompson** [comment, marked as answer]
Hrm, this is an interesting question. So I'll be clear, the library wasn't designed with async/await in mind at all. It'd be possible to add closures to steps that'd be async/await compatible, for that matter it'd be good if the existing closures would be updated to support throwing errors as well. 

The design for this may take me a bit to nail down. I can't give you great advice on making this work right this second but I'll tell you that CucumberSwift steps execute synchronously, It might be possible for you to build something yourself that utilizes XCTWaiter() and an async closure to kind of hack in the functionality. 

I'm happy to add it to the backlog of ideas, though. It's a totally reasonable feature request.

**Reply by fwpascual** [reply]
Thanks for your reply.  I think adding async/await would be a nice addition along with supporting throwing errors.  If you were to add this to the backlog, that would be great.  I can certainly work around these issues in the meantime.  I will look at `XCTWaiter` as a workaround for now.
